Multi-core build is failing.

Steps to reproduce: 
1) change "num_cores" in champsim_config.json -->  "num_cores": 2
2) ./config.sh champsim_config.json
3) make

src/core_inst.cc:28:8: error: redefinition of ‘O3_CPU cpu0’
   28 | O3_CPU cpu0(0, 1.0, 32, 8, 16, 64, 32, 32, 352, 128, 72, 6, 6, 6, 128, 4, 2, 2, 5, 1, 1, 1, 0, 0, &cpu0_ITLB, &cpu0_DTLB, &cpu0_L1I, &cpu0_L1D, O3_CPU::bpred_t::bbranchDbimodal, O3_CPU::btb_t::bbtbDbasic_btb, O3_CPU::ipref_t::pprefetcherDno_instr);

src/core_inst.cc:27:8: note: ‘O3_CPU cpu0’ previously declared here
   27 | O3_CPU cpu0(0, 1.0, 32, 8, 16, 64, 32, 32, 352, 128, 72, 6, 6, 6, 128, 4, 2, 2, 5, 1, 1, 1, 0, 0, &cpu0_ITLB, &cpu0_DTLB, &cpu0_L1I, &cpu0_L1D, O3_CPU::bpred_t::bbranchDbimodal, O3_CPU::btb_t::bbtbDbasic_btb, O3_CPU::ipref_t::pprefetcherDno_instr);
      |        ^~~~
make: *** [<builtin>: src/core_inst.o] Error 1

--- a/.github/workflows/buildcheck.yml
+++ b/.github/workflows/buildcheck.yml
@@ -47,7 +47,7 @@ jobs:
         name: champsim-singlecore-results
         path: single_core_result.txt
 
-  single_core:
+  multi_core:
     runs-on: ubuntu-latest
 
     steps:

--- a/.github/workflows/buildcheck.yml
+++ b/.github/workflows/buildcheck.yml
@@ -24,7 +24,7 @@ on:
       - '!.github/workflows/buildcheck.yml'
       
 jobs:
-  build:
+  single_core:
     runs-on: ubuntu-latest
 
     steps:
@@ -38,11 +38,34 @@ jobs:
       
     # TODO run an actual set of checks
     - name: Run tests
-      run: ./bin/champsim -w50000000 -i50000000 -- https://hpca23.cse.tamu.edu/champsim-traces/speccpu/400.perlbench-41B.champsimtrace.xz > result.txt
+      run: ./bin/champsim -w50000000 -i50000000 -- https://hpca23.cse.tamu.edu/champsim-traces/speccpu/400.perlbench-41B.champsimtrace.xz > single_core_result.txt
 
     # We upload the generated files under github actions assets
     - name: Upload Results
       uses: actions/upload-artifact@v2
       with:
-        name: champsim-results
-        path: result.txt
+        name: champsim-singlecore-results
+        path: single_core_result.txt
+
+  single_core:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2.3.4
+
+    - name: Configure
+      run: ./config.sh <(sed 's/"num_cores": .*,/"num_cores": 2,/' champsim_config.json)
+      
+    - name: Make
+      run: make
+      
+    # TODO run an actual set of checks
+    - name: Run tests
+      run: ./bin/champsim -w50000000 -i50000000 -- https://hpca23.cse.tamu.edu/champsim-traces/speccpu/400.perlbench-41B.champsimtrace.xz https://hpca23.cse.tamu.edu/champsim-traces/speccpu/401.bzip2-226B.champsimtrace.xz > multi_core_result.txt
+
+    # We upload the generated files under github actions assets
+    - name: Upload Results
+      uses: actions/upload-artifact@v2
+      with:
+        name: champsim-multicore-results
+        path: multi_core_result.txt

--- a/config.sh
+++ b/config.sh
@@ -299,7 +299,7 @@ for elem in memory_system:
             pref_enum_string=' | '.join(f'CACHE::p{k}' for k in elem['prefetcher']),\
             **elem)
 
-instantiation_file += ',\n'.join(
+instantiation_file += ''.join(
         'O3_CPU ' + cpu['name'] + cpu_fmtstr.format(
             branch_enum_string=' | '.join(f'O3_CPU::b{k}' for k in cpu['branch_predictor']),
             btb_enum_string=' | '.join(f'O3_CPU::t{k}' for k in cpu['btb']),

--- a/config.sh
+++ b/config.sh
@@ -80,11 +80,6 @@ cores = config_file.get('ooo_cpu', [{}])
 # Index the cache array by names
 caches = {c['name']: c for c in config_file.get('cache',[])}
 
-# Default branch predictor and BTB
-for i in range(len(cores)):
-    cores[i] = ChainMap(cores[i], {'name': 'cpu'+str(i), 'index': i}, copy.deepcopy(dict((k,v) for k,v in config_file.items() if k not in ('ooo_cpu', 'cache'))), default_core.copy())
-    cores[i]['DIB'] = ChainMap(cores[i]['DIB'], config_file['DIB'].copy(), default_dib.copy())
-
 # Copy or trim cores as necessary to fill out the specified number of cores
 original_size = len(cores)
 if original_size <= config_file['num_cores']:
@@ -92,6 +87,11 @@ if original_size <= config_file['num_cores']:
         cores.append(copy.deepcopy(cores[(i-1) % original_size]))
 else:
     cores = cores[:(config_file['num_cores'] - original_size)]
+
+# Default branch predictor and BTB
+for i in range(len(cores)):
+    cores[i] = ChainMap(cores[i], {'name': 'cpu'+str(i), 'index': i}, copy.deepcopy(dict((k,v) for k,v in config_file.items() if k not in ('ooo_cpu', 'cache'))), default_core.copy())
+    cores[i]['DIB'] = ChainMap(cores[i]['DIB'], config_file['DIB'].copy(), default_dib.copy())
 
 # Append LLC to cache array
 # LLC operates at maximum freqency of cores, if not already specified

